### PR TITLE
Added Roman Numerals for easy conversion when comparing titles with some refactorings in MovieRepository 

### PR DESCRIPTION
**Pull request by Rusk85**
#### Database Migration
NO
#### Description
(_Edit: Was in a bit of a hurry earlier when writing the PR comment, so having done a rewrite_)
## Concrete Changes
1. Added new .NET4 Library Project [RomanNumerals](https://github.com/Rusk85/RomanNumeral) which is originally from this repository [adamfisher/RomanNumeral](https://github.com/adamfisher/RomanNumeral) and supports conversion of any `int` in the range `1-3999`
2. Added new project as dependency to `NzbDrone.Core` for usage in `MovieRepository` when calling `FindByTitle(string cleanTitle, int? year)` for evaluating if a movie is part of a franchise
3. Moved all functionality of `MovieRespository.FindByTitle(string cleanTitle)` and `MovieRespository.FindByTitle(string cleanTitle, int year)` to `private Movie MovieRespository.FindByTitle(string cleanTitle, int? year)`
4. While moving said functionality in _3._ I refactored both methods with the DRY principle in mind because both public methods had a lot of overlap
5. I used the `RomanNumerals` to create a `static Dictionary<string,string>` that maps all Arabic Numeral values in the range of `1 .. __supportedNumberOfMovieParts +1` (currently hardcoded to 20) and thread-safely initialize the Dictionary once (per Request/User session/Globally => from ASP.NET I know that using `static` is the equivalent of an variable with the scope of the whole Appliation in that particular IIS AppPool but don't know how that's handled here)

## Some more information
- I saw after opening my PR that you run a CI pipeline. I assume its configured to run the tests as well but only those relevant to radarr I assume? Do you use [SonarQube](https://www.sonarqube.org/) as a code quality gate perhaps? There are plugins for jenkins for it out there as well.
- As for the `RomanNumerals` project I had to create an additional project within that solution mentioned in _1._ and target .NET 4 since changing the target in the existing project was not possible. From there on I really just copy & pasted the 1 (!) class that's doing everything. Didn't have to change a single character of the logic in that class

## How I decided for this particular change
I was familiarizing myself with the code trying to figure out flow of control/data and such and kind of stumbled over this. Saw it as a worthwhile chance to get some hands-on experience with the code base

#### Todos
- Tests required for new submodule `RomanNumerals` (original author has tests in place which could be taken as a good start) as well as the refactored part in `MovieRepository`
- Think about making `_supportedNumberOfMovieParts` in `MovieRepository` part of some overall config since its hardcoded; maybe even adapt to be completely dynamic 
- Since `RomanNumeral` really just consists of one class it might be worthwhile thinking about merging it into the code tree and extract the tests from the existing test project
#### Issues Fixed or Closed by this PR
- None

### CODE REVIEW

**Comment by onedr0p** [comment]
1) Your thoughts, could we include that RomanNumerals.cs library in our project instead of a external submodule?

2) You will notice RomanNumerals in other places too:

https://github.com/Radarr/Radarr/blob/b81d8c4678acfa6d655366c2edbc1a91f85dbd31/src/NzbDrone.Core/Parser/ParsingService.cs#L36

**Comment by onedr0p** [comment]
Feel free to join discord if you would like to chat.

**Comment by galli-leo** [comment]
Also I think 20 is a fine value (As you can take from my previous comment ;)).

**Comment by galli-leo** [comment]
Additionally, seems good to merge.

**Comment by onedr0p** [comment]
Waiting on answers for my questions above though.

**Comment by Rusk85** [comment]
Sorry got distracted mid-answer. Writing one up right now. Give me 10 minutes.

**Comment by Rusk85** [comment]
> 1. Your thoughts, could we include that RomanNumerals.cs library in our project instead of a external submodule?
Yes when writing up the PR description it occurred to me that we could just as easily merge the `RomanNumerals`project into the internal code tree of radarr since RomanNumerals exists litteraly out one file/class. Whats more the author of the original repo put it up under MIT. And I am assuming we are probably looking at a bit of a chaotic landscape when it comes to tests. So tossing in one more test class for the numerals wouldn't hurt. I mean the tests are virtually there written by the original author and will probably need some good old copy&paste to solve the projects targeting issue and then sprinkle it w/ w/e annotations the testing framework used in radarr requires.

> 2. You will notice RomanNumerals in other places too:
Never occurred to me to look at other places that would require this. My first impulse would be to first look for any further places where Roman numerals are converted to Arabic numerals or even just incidents where something is done with Roman numerals that could be satisfied by the project `RomanNumerals`. The next step would be to then extract what code I have written for the conversion and put it in its own class so I can be called upon from multiple places in radarr's code base. As for the current PR I would just go ahead and pull the trigger. Collecting the necessary information required to inform the next steps in this  matter are unavoidable anyway. And I am pretty sure from glancing at that one spot you pointed out that extraction of my specialised conversion logic has to happen anyway. If I do that own my own branch from where the PR originates or can do a `git pull && git checkout develop && git checkout -b 0000_extract_roman-numeral-conversion_behavior` doesn't really matter to me.

>20?
Well if you want to do this fully dynamically that is keeping most parts of how potential movie parts are being identified and determine the required range to run through than my first thought would be some non-trivial regex but if you have arrived at that point then why keep the loop at all and not just take your regex, extract the roman numeral, convert to arabic numeral and relpace in `cleanTitle'. I arrived at the number 20 by simply doubling what the previous limit was. If you want to have more control over this without having to necessarily change the code than you are looking at some form of a serialized piece of information holding the range of movie parts you would like to support. Possibly you want to be able to change that value using the API or even expose that implementation detail to the user for him to make a choice.

Right did I answer everything. All good?

**Comment by onedr0p** [comment]
All good, I can merge the PR unless you want to update the Test, ParsingService, and bring in the class. I feel as though we can wait until those things get done because it really doesn't fix anything.

**Comment by Rusk85** [comment]
Alright lets do it like this then:
- I'll merge RomanNumerals into the internal code tree
- add the tests for it
- create a ParsingService and wire that up to all places that require the same functionality

**Comment by onedr0p** [comment]
Awesome, thanks :)

**Comment by Rusk85** [comment]
Just wanted to say I haven't forgotten this but I am kinda tied up at work. Working buddy needs me to integrate/rewrite a component he needs by Monday. I'll try to dedicate some time to this PR later in the evening.You can follow my progress on this branch in my fork: [1217_PR_merge-romans-internal_add-romans-test_add-parsing-service](https://github.com/Rusk85/Radarr/tree/1217_PR_merge-romans-internal_add-romans-test_add-parsing-service)

**Comment by Rusk85** [comment]
While I am working on the things we have agreed upon I am also trying to stay in sync with upstream commits so I am pulling all the changes from this repo in my fork, merging them their into my local develop and from there on out merging downwards to a) the branch from my fork I used to open this PR and b) the new branch I have opened to address the things left for this PR to be accepted with the goal of `git checkout pr_add_roman_numerals_converter && git merge 1217_PR_merge-romans-internal_add-romans-test_add-parsing-service && git push origin`. That should work in the end right?

**Comment by Rusk85** [comment]
Ok 1 Test out of ca 2.4k failed with this error message:
    
    System.IO.DirectoryNotFoundException : Could not find a part of the path 'C:\projects\radarr-usby1\_tests\_temp_636263432504931207\crk2u30t.w5c'.

Probably pulled that in when trying to keep up with upstream. Will have to investigate further tomorrow.

File: src/NzbDrone.Core/NetImport/NetImportSearchService.cs [hunk 1]
```diff
@@ -81,19 +81,7 @@ public List<Movie> MovieListSearch(int listId, bool onlyEnableAuto = false)
...
-					return x.TmdbId.ToString();
-				}
-
-				if (x.ImdbId.IsNotNullOrWhiteSpace())
-				{
-					return x.ImdbId;
-				}
-
-				return x.Title;
-			}).ToList();
```
Comment (onedr0p) [inline src/NzbDrone.Core/NetImport/NetImportSearchService.cs]: I think we want to keep those changes... careful with merging.

File: src/NzbDrone.Core/Parser/ParsingService.cs [hunk 2]
```diff
@@ -1,668 +1,682 @@
...
+        private Movie GetMovie(ParsedMovieInfo parsedMovieInfo, string imdbId, SearchCriteriaBase searchCriteria)
+        {
+            Movie movieByCriteria;
+            if ((movieByCriteria = TryFindMovieBySearchCriteria(parsedMovieInfo, searchCriteria)) != null)
+            {
+                return movieByCriteria;
+            }
+
+            Movie movieByTitle;
+            if ((movieByTitle = TryFindMovieByTitle(parsedMovieInfo, searchCriteria)) != null)
```
Comment (onedr0p) [inline src/NzbDrone.Core/Parser/ParsingService.cs]: I like how you split those into separate methods. Make it more clear!

File: src/NzbDrone.Core/NetImport/NetImportSearchService.cs (hunk 1 above)
Comment (Rusk85) [inline src/NzbDrone.Core/NetImport/NetImportSearchService.cs]: Totally sorry for the merge f'ups but looked for the first time of my life at a kdiff3 or some such with 3 panels. Think I kinda figured it out. "Theirs" is the branch you are merging from, "Ours" the branch we are merging into and the middle screen is basically the result.

File: src/NzbDrone.Core/Parser/ParsingService.cs (hunk 2 above)
Comment (Rusk85) [inline src/NzbDrone.Core/Parser/ParsingService.cs]: Yup hate super long methods with like one local variable that is abused throughout the whole 400 lines of method scope for all kinds of nasty things. I must admit though I am using ReSharper for  this kind of breaking methods up. Just a matter of setting your cursor to the start of a large statement block and then use ReSharper 's right extend and extract methods functionality. Granted afterwards you have to do some manual adjust for return types or parameter passing with ref or out. And yeah it doesn't hurt to have a basic understanding of what is going on with the parts of the code you are looking to refactor.

**Comment by Rusk85** [comment]
Seeing that I have a hard time keeping up with upstream in my fork I would really like to go ahead with the PR now. Had to do a 3-way merge on multiple files for the second day in a row. That is a total time sink for me because I do get the basics of git and try to do as much as I can using the cmd  line I am easily occupied for up to 2hrs figuring out what goes where when merging/rebasing. I have added the Roman Numerals along with a parser with a super easy to use API. Also removed at least one spot where a dictionary for mapping Arabic to Roman numerals was used. So only really missing at this point would be tests and potential other spots that use manual numeral mappings.
Cuz once the PR goes through I can throwaway my fork and refork and start with a blank slate (finally) so to say.
So what say you oh ppl wielding the mighty power of write access?  😃

File: src/NzbDrone.Core/Tv/MovieRepository.cs [hunk 3]
```diff
@@ -277,6 +187,47 @@ public PagingSpec<Movie> MoviesWhereCutoffUnmet(PagingSpec<Movie> pagingSpec, Li
             return pagingSpec;
         }
 
+        private Movie FindByTitle(string cleanTitle, int? year)
+        {
+            cleanTitle = cleanTitle.ToLowerInvariant();
+            string cleanTitleWithRomanNumbers = cleanTitle;
+            string cleanTitleWithArabicNumbers = cleanTitle;
+
+            throw new NotImplementedException("Complete work on branch 1217_PR_merge-romans-internal_add-romans-test_add-parsing-service");
```
Comment (Rusk85) [inline src/NzbDrone.Core/Tv/MovieRepository.cs]: That line has to be removed! I'll.do it once back.from.work

**Comment by onedr0p** [comment]
Let me know when the PR is good to merge.

**Comment by Rusk85** [comment]
Sounds good. Just came home from work. Fixing up the last few bits in the PR and will give you the go-ahead during the next couple of hours depending on how many more pushes in to develop were made since yesterday and if any generate conflicts when merging into my local develop/add_roman_numerals branches. Fixing everything up shouldn't take very long. Anyway commencing work **now**.

**Comment by Rusk85** [comment]
**Okay I officially surrender!**
(_tl;dr at the bottom_)

I have destroyed fork on github and especially my local repo beyond recognition. Like doing rebases including up to 14 steps starring at 3 panels all showing different versions of the same file w/o having a clue what comes from where and seemingly from one `git rebase --continue` to the next then changes I made in one panel appeared in another one. 
Well trying to maintain two remotes (this repo I named `upstream` and my github fork named `origin`) while working on one file that seems very well frequented was just too much for my limited knowledge about merges and rebases (which I didn't ask for since I was just pulling from your this repo) wasn't a good idea. 
Add to that the fact that my workflow wasn't very consistent. My plan was just to pull changes from `upstream/develop` into my local `develop` branch, push those up to `origin/develop` and then merge `develop` into my local `add_roman_numerals`. 
But then I created a branch of off that one `1217_PR...` and merge that one into `add_roman_numerals`. Might have still been fine but I think at some point I even merged those branches into my local `develop` and the other way around. 
So now its in a total fubar state and I can't even recognize the few refactorings to an existing file after the continuous `git rebase` operations I did over the course of a week or so.

But as broken as my current repo is, that doesn't really matter that much. In essence I have added a new folder somewhere in `NzbDrone.Core` where I maintained all files relating to Roman numerals (conversion). Additionally I have made refactorings to 2 files which are still in a working state further down the history. So going forward I will 
- close this PR
- delete my fork of this repo on gitub
- then re-create it
- add my files to `NzbDrone.Core`
- re-apply my refactorings using the working version I will checkout from my local repo of my then deleted `origin`
- finally create a new PR

# tl;dr
My local repo is completely messed up due to trying to keep up with changes on files I was actively working on and mistakes on my part regarding my git workflow. Since I can easily isolate the new feature I added (Bi-Directional conversions between Roman und Arabic numerals) as well as re-apply my limited refactorings to the latest versions of the files, I will close this PR, delete and re-create my fork on github of this repo and once done with my changes create a new PR.

### All this of course assuming that you are still interested in my contributions after this exercise in grandiose failure.

**Comment by onedr0p** [comment]
No worries, it's a learning experience :) I find it hard to maintain forks as well.

**Comment by Rusk85** [comment]
Glad to hear you are seeing it that way. Alright closing this PR and putting into effect **Plan B**. ^^
